Tim 24D
* SC62404 - Frasty Julistyfenia Rambing
* SC62410 - Siti Luna Satiyah
* SC62418 - Winola Farhani
* SC62419 - Anindyta Shayna Widyanugraha
* SC62421 - Hastialisna Hurul Aeni Setiawan
* SC62423 - Perwira Annissa Dyah Permatasari



Text classification berdasarkan Bag-of-Words

Tulis syntax di Google Colab:
- Tentukan tiga topik terkait: Kesehatan, Olahraga, Keuangan.
- Temukan teks/artikel sebagai dasar/pelatihan/training ketiga topik tersebut.
- Hitung vektor Bag-of-Words untuk setiap topik.

Klasifikasikan teks ini dalam topik yang benar:\
“Cristiano Ronaldo came off the bench to earn Manchester United a hard-fought 2-1 victory at Everton in the Premier League on Sunday, taking his career goal tally to 700 in the process. Just as United did last weekend in their derby mauling at the hands of local rivals Manchester City, they again found themselves behind early on at Goodison Park after Alex Iwobi curled a sublime strike into the net from 20 metres.”

#### Topik dan Artikel

**Topik**:
* Topik kesehatan: Mental Health & Medical Act
* Topik olahraga: Basketball & Football
* Topik keuangan: Tax & Indonesia's Economic

**Artikel dan Sumber:**\
Training article
* Mental Health : [A promising framework for managing students’ mental health](https://www.universityworldnews.com/post.php?story=20240415151821371)
* Basketball : [Day 1: Prawira overcome adversity, rally back to beat Hong Kong Eastern](https://www.fiba.basketball/bclasia/2024/qualifiers/news/round-2-day-1-recap)
* Tax : [President Joe Biden Proposes Historic 44.6% Capital Gains Tax](https://coingape.com/president-joe-biden-proposes-historic-44-6-capital-gains-tax/)

Testing article
* Medical Act : [MOH Stands by Decision to Amend Medical Act 1971](https://thesun.my/local_news/moh-stands-by-decision-to-amend-medical-act-1971-GI12388295)
* Football : [Indonesia U-23 National Team Makes History by Beating South Korea in Penalty Shootout](https://jakartaglobe.id/news/indonesia-u23-national-team-makes-history-by-beating-south-korea-in-penalty-shootout)
* Indonesia's Economecic : [Finance Minister Forecasts Indonesia's Q1 2024 Economic Growth at 5.17 Percent](https://jakartaglobe.id/business/finance-minister-forecasts-indonesias-q1-2024-economic-growth-at-517-percent)

#### Menghitung Vektor Bag-of-Words

##### Import Library

In [ ]:
# Library
import nltk
nltk.download("popular")
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import CountVectorizer
from collections import Counter

[nltk_data] Downloading collection 'popular'
[nltk_data]    | 
[nltk_data]    | Downloading package cmudict to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cmudict.zip.
[nltk_data]    | Downloading package gazetteers to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/gazetteers.zip.
[nltk_data]    | Downloading package genesis to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/genesis.zip.
[nltk_data]    | Downloading package gutenberg to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/gutenberg.zip.
[nltk_data]    | Downloading package inaugural to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/inaugural.zip.
[nltk_data]    | Downloading package movie_reviews to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/movie_reviews.zip.
[nltk_data]    | Downloading package names to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/names.zip.
[nltk_data]    | Downloading package shakespeare to /root/nltk_data...
[nlt

##### Training Article

In [ ]:
# Article for training set
medical = "Sharing the responsibility for mental health support across an institution can reduce the burden on a university counselling centre. Yale-NUS College, a residential liberal arts institution jointly established by Yale University and the National University of Singapore, utilises trained student affairs staff and academic advisors to triage and manage lower-level mental health issues including mild forms of academic stress and social anxiety.  Staff complete in-house training organised by counselling colleagues and some extend their training through external programmes such as Mental Health First Aid. In 2021, as an increasing number of teaching faculty encountered mental health struggles among their students, Yale-NUS introduced ‘gatekeeper training’ that covered how to recognise signs of emotional distress, taught empathic and active listening skills and explained the various resources available to students. Faculty and teaching assistants are often the first to recognise student mental health issues since they engage with students regularly in the classroom, at advising appointments and during office hours. Administrators should ensure faculty have access to and utilise an internal referral system that alerts the appropriate staff when professional follow-up is necessary. NUS understood the importance of involving multiple stakeholders when they created the WellNUS© Mental Health Framework in 2021 to systematically map out the different aspects of wellbeing and identify the relevant initiatives and key stakeholders that can provide support. The aim of the framework is to have a more holistic, structured and sustainable approach towards student and staff wellbeing. In 2021, Yale-NUS introduced a for-credit “Resilience and Success in College” elective course for first-year students in order to equip them with strategies to respond to the inevitable challenges and adversity that they will face during college.  Example topics include finding purpose, utilising strengths, practising vulnerability and developing positive interpersonal relationships. The course was later adapted for graduating seniors to help support them through the often-stressful phase of transitioning out of college. The six-week course was a collaborative effort between the student affairs division and the psychology faculty with seminar-style classroom discussions and assignments that emphasised personal reflection and application. Students who completed the six-week course reported that they were better equipped to navigate future issues and understand what support resources are available to them."
sport = "Prawira Harum Bandung overcame adversity and stunned Hong Kong Eastern Basketball Team, 53-50, for a triumphant start to their Round 2 campaign in the Basketball Champions League Asia (BCL Asia) Qualifiers, Tuesday night at the BritAma Arena. The reigning Indonesian Basketball League (IBL) champions lost star scorer Brandone Francis just 15 seconds into the third quarter due to a disqualifying foul. However, that became an opportunity for the whole crew to play collectively and soon get back on their Round 1 tormentors. Prawira went on an 18-5 scoring tear in that whole quarter to head into the final period ahead, 44-38, and momentum never left their side the rest of the way en route to joining another IBL outfit in Pelita Jaya Basketball as the Round 2 opening night winners. Antonio Hester top-scored for the hometown side with 19 points and 15 rebounds. James Gist fell just short of a double-double outing with 9 points and 13 rebounds, together with 3 assists, 4 steals, and 3 blocks on his way to netting an efficiency (EFF) of 21. Indra Muhamad had 8 points while Yudha Saputera added 7 points as the spitfire guards nailed two three-pointers apiece to help out in the third-quarter reversal. Saputera nailed a trey to tie things up and Muhamad followed it up with a triple of his own to give themselves the 37-34 lead with 4:13 left, putting them ahead for good, Prawira would lead by as many as nine, 52-43, late in the game behind a jumper by Gist. Now they look forward to their meeting against Pelita Jaya on Wednesday, April 24, 20:00 local time, with the winner virtually assured of a spot in the BCL Asia proper. Chris McLaughlin, meanwhile, led Hong Kong Eastern with 17 points and 10 rebounds. Yiu Pong Yip was the only other double-digit scorer with 12 in the collapse. The reigning Hong Kong A1 Division champions still tried to come back by coming within just a field goal, 52-50, following Dominic Gilbert's conversion with 35.4 seconds left, but a split by Saputera plus Prawira's lockdown defense prevented them from coming back. Hong Kong Eastern will try to bounce back against the NS Matrix Deers on Wednesday as well at 17:30 local time in a battle of teams coming off losses."
finance = "United States President Joe Biden has proposed to raise capital gains tax to about 44.6%, marking the highest formal federal capital tax in history. Per a Forbes report, this proposal was added to Biden’s budget proposal for the 2025 fiscal year. A footnote from the General Explanations of the Administration’s Fiscal Year 2025 Revenue Proposals reads “A separate proposal would first raise the top ordinary rate to 39.6 percent … An additional proposal would increase the net investment income tax rate by 1.2 percentage points above $400,000 … Together, the proposals would increase the top marginal rate on long-term capital gains and qualified dividends to 44.6 percent.” There is a main proposal which gives context to the above-mentioned statement and this involves raising the long-term capital gains and qualified dividends rates to 37% for taxpayers. This is for taxpayers with taxable income above $1 million. The potential of having the 44.6% rate come to fruition is only possible under a different proposal from the president’s administration’s main capital gains rate increase. In like manner, this rate would apply to only those who have taxable incomes of $1 million and $400,000 in investment income. Presenting such a capital gains rate proposal is a subtle policy maneuver that will likely trigger a high percentage while ignoring the crucial aspect of income thresholds. It appears that the policy is attempting to place high ordinary income and investment income earners on a common playing field. Amid the administration’s push to hike taxes, many of the businesses and individuals that may be impacted may turn to digital assets to help them experience economic freedom. At least, the laws of crypto tax reporting have not fully taken effect, nor does it require a tax rate as high as what Biden is proposing. The U.S. Internal Revenue Service (IRS) revealed an early draft of a new tax form for reporting crypto transactions a few days ago. The form, dubbed 1099-DA, is aimed at facilitating and simplifying tax liabilities linked to crypto transactions. It takes a record of taxable gains or losses and contains some sections to mention specific token codes, wallet addresses, and other transaction information. Many investors are likely to find the form easy to fill, considering that some tax experts took time out to explain the appropriate approach to do so."

# Article for test set
medical_test = "The Ministry of Health (MOH) stands by its decision to amend the Medical Act 1971 (Act 50) to bridge the gap between the parallel pathway programme (PPP) and the local medical degree programme, despite objections from some parties. Health Minister Datuk Seri Dr. Dzulkefly Ahmad said after conducting a comprehensive engagement session with the Ministry of Higher Education (MOHE), it was found necessary to make amendments to the act. He said the ministry has also received legal views from the Attorney General's Chambers. “Anyone can express their opinion, but both MOH and MOHE are aligned on this matter, in line with the comprehensive joint study and engagement. Therefore, we will soon submit this proposed amendment memorandum to the Cabinet. “This is the process we are pursuing to resolve the issue, to clarify the status of both programmes and to address the shortage of medical professionals serving in government healthcare facilities across the country,“ he told a press conference after a working visit to Kuala Kubu Baharu (KKB) Hospital here today. In early April, Dr Dzulkefly announced that the MOH would propose amendments to Act 50 to ensure prompt registration and recognition of medical professionals trained under the PPP, no later than the Second Meeting of the Third Session of the 15th Parliament in June this year. However, the Group of Professors of Health and Medicine, led by its chairman, Prof Dr. Noor Hassim Ismail, disagreed, stating that the gap between the PPP and the local medical degree programme could be resolved without amending Act 50. Commenting on his visit to KKB Hospital today, Dr Dzulkefly said improvements in healthcare services, such as the placement of general medical practitioners since January last year, have successfully reduced congestion at Selayang Hospital. “Previously, there were no specialists here, so KKB residents had to go to Selayang Hospital for treatment. Now, with the presence of medical specialists, I see that we have gained their confidence to seek treatment here,“ he said. “Not only that, to enhance healthcare services at KKB, we plan to establish an emergency department (currently only an emergency unit) staffed with emergency medicine specialists by early May,“ he said. Dr Dzulkefly said other aspects needing improvement in government healthcare facilities in KKB include enhancing services for the surrounding population which includes addressing congestion in the pharmacy unit and providing additional parking spaces."
sport_test = "The Indonesia U-23 national team made history by defeating South Korea in a penalty shootout during the quarter-finals of the U-23 Asian Cup. Indonesia emerged victorious in the penalty shootout, winning 11-10 after the match ended 2-2 following 120 minutes of play. This victory secured Indonesia's place in the semifinals. The match, held at the Abdullah bin Khalifa Stadium in Doha on Thursday (early Friday in Jakarta time), saw South Korea initially breach Indonesia's goal in the 8th minute. However, after a VAR review, the goal was disallowed by the referee due to the South Korean player being in an offside position. Facing the Asian powerhouse, Indonesia broke the deadlock in the 15th minute. Rafael Struick's shot from outside the penalty box found the bottom left corner of the South Korean goal, beyond the reach of goalkeeper Baek Jong-bum. The one-goal advantage boosted Indonesia's confidence. Marselino created a good opportunity after combining with Witan Sulaeman, but unfortunately, Marselino's shot sailed over the South Korean crossbar. South Korea equalized in the 45th minute through a header that deflected off Komang Teguh and into Ernando Ari's goal. Just before the end of the first half, Struick restored Indonesia's lead. Receiving a long pass from Ivar Jenner, he controlled the ball, deceived the opponent's defender, and then beat goalkeeper Baek with a left-footed shot. The first half ended with Indonesia leading 2-1. Early in the second half, Indonesia had three good chances, but two shots from Struick and one from Pratama Arhan missed the target. Marselino had another good chance when dribbling on the left side, but his shot found only the side netting of the South Korean goal. Shortly afterward, coach Shin Tae-yong made the first substitution, replacing Rio Fahmi with Fajar Fathurrahman. In the 68th minute, Lee Young-jun committed a harsh foul on Hubner."
finance_test = "Finance Minister Sri Mulyani Indrawati predicts Indonesia's economic growth in the first quarter of 2024 will reach 5.17 percent, driven by strong household consumption. However, this growth is still constrained by the performance of exports and imports, which often face contractions due to the impact of global economic turmoil. 'Supported by strong performance from consumers, manufacturing, as well as foreign direct investment (FDI) and the purchasing managers index (PMI), we predict our growth for Q1 2024 to be at 5.17 percent. However, we remain cautious about the turbulence occurring in the global economy,' said Sri Mulyani during a press conference at her office on Friday. Government consumption is expected to grow strongly in the first quarter driven by election spending, salary increases for civil servants, and the distribution of holiday bonuses (THR) in the first quarter. Household consumption also increased significantly in line with election activities, THR spending, and controlled inflation. The Consumer Confidence Index (CCI) in March 2024 was recorded at 123.8, indicating continued consumption growth. While electricity consumption for businesses remains positive, there was a contraction in the industrial sector. Sri Mulyani attributed this contraction to the numerous public holidays before Eid al-Fitr. Cement consumption, which surged two months ago, corrected to 1.9 percent growth in March. 'Overall, consumer spending is quite good, but we need to remain cautious because some sectors experienced corrections, both seasonal, such as Ramadan and Eid, and structural long-term corrections,' said Sri Mulyani. Investment is expected to increase alongside the progress of national strategic projects (PSN) and the development of the new capital city, Nusantara Capital (IKN). Meanwhile, export and import activities are still hindered by weak global demand. Indonesia's trade balance recorded a surplus of $4.47 billion (Rp 72.45 trillion) in March 2024, up from $3.64 billion in February 2024. This marked a significant increase from March 2023, which stood at $1.65 billion. March 2024 exports reached $22.43 billion, marking a 16.4 percent increase from February 2024, but a 4.19 percent contraction compared to March 2023. Meanwhile, imports in March 2024 totaled $17.96 billion, reflecting a 2.6 percent decrease from February 2024 and a 12.76 percent decrease from February 2023."

texts = [medical, sport, finance, medical_test, sport_test, finance_test]
bow_keys = []
corpus_texts = []

lemmatizer = WordNetLemmatizer()
for text in texts:
  words = word_tokenize(text) # Tokenize the text into individual words
  texts = [lemmatizer.lemmatize(word) for word in words if word not in set(stopwords.words('english'))] # Lemmatize each word and filter out English stopwords
  bow_keys += texts # Store the lemmatized words
  text = ' '.join(texts) # Join the lemmatized words back into a single string for each text
  corpus_texts.append(text)

bow_keys = set(bow_keys) # Update the bow_keys list

# Print result
print(bow_keys)
print(corpus_texts)

{'13', 'Thursday', 'raising', 'act', 'FDI', 'aim', 'considering', 'indicating', "'", 'reported', 'look', 'process', 'crew', 'Nusantara', 'Not', 'breach', 'broke', '400,000', 'systematically', 'pathway', 'complete', 'second', 'outfit', 'outing', 'unfortunately', 'medicine', 'added', 'investor', 'net', 'fully', 'Basketball', 'seasonal', 'providing', 'Pratama', 'Ministry', 'Proposals', 'facility', 'coming', 'stating', 'taken', 'Datuk', 'winning', 'U-23', 'Asian', 'prompt', 'predict', 'Gilbert', 'adapted', 'holiday', 'team', 'hike', 'course', 'Presenting', 'come', 'session', 'utilising', 'despite', 'resource', 'became', 'capital', 'export', 'industrial', '1099-DA', 'Ahmad', 'subtle', 'The', 'provide', '10', 'among', 'EFF', '22.43', 'Khalifa', 'Kong', 'thing', 'boosted', 'early', 'reach', 'household', 'teaching', 'still', 'relevant', 'Service', 'Investment', 'Confidence', '45th', 'country', 'Biden', 'bonus', 'manufacturing', 'Shin', 'appointment', 'responsibility', 'continued', 'night', 'la

Dalam loop for, terdapat perlakuan sebagai berikut:

1. Setiap teks dari texts di-tokenisasi menjadi kata-kata menggunakan word_tokenize.

2. Kemudian, dilakukan lemmatisasi untuk setiap kata dalam teks menggunakan lemmatizer.lemmatize(word) untuk mengubah kata-kata ke bentuk dasar.
Kata-kata yang bukan stop words (kata-kata umum yang tidak memberikan banyak informasi seperti "the", "is", "and", dsb.) disimpan dalam variabel texts.

3. Setiap kata yang sudah diproses ditambahkan ke list bow_keys yang merupakan bag-of-words keys.

4. Teks yang sudah diproses dan digabungkan kembali diubah menjadi teks utuh dan ditambahkan ke list corpus_texts.

Kemudian setelah selesai melakukan preprocessing, pada set(bow_keys) digunakan untuk menghilangkan kata-kata duplikat dari bag-of-words keys. Hasil bag-of-words keys yang sudah dibersihkan ditampilkan menggunakan print(bow_keys) dan hasil teks yang sudah diproses digabungkan kembali (cleaned data) ditampilkan menggunakan print(corpus_texts).

##### Vektor Bag-of-Words

In [ ]:
# Initialize a dictionary to store BoW vectors for each topic
bow_vectors = {}

# Iterate over each topic's processed text corpus
for idx, topic_text in enumerate(corpus_texts):
    bow_vector = Counter() # Initialize a Counter to count word occurrences
    topic_words = topic_text.split() # Tokenize the processed text

    # Count occurrences of each word in the BoW keys
    for word in topic_words:
        if word in bow_keys:
            bow_vector[word] += 1
    bow_vectors[f"Topic {idx+1}"] = bow_vector # Store the BoW vector for the topic in the dictionary

# Print the BoW vectors for each topic
for topic, bow_vector in bow_vectors.items():
    print(topic)
    print(bow_vector)
    print()

Topic 1
Counter({'.': 13, ',': 10, 'student': 8, 'mental': 4, 'health': 4, 'support': 4, 'course': 4, 'Yale-NUS': 3, 'staff': 3, 'issue': 3, 'training': 3, '2021': 3, 'faculty': 3, 'The': 3, 'institution': 2, 'counselling': 2, 'College': 2, 'University': 2, 'affair': 2, 'academic': 2, 'Mental': 2, 'Health': 2, 'In': 2, 'teaching': 2, 'introduced': 2, 'recognise': 2, 'resource': 2, 'available': 2, 'classroom': 2, 'stakeholder': 2, 'wellbeing': 2, 'college': 2, 'six-week': 2, 'Sharing': 1, 'responsibility': 1, 'across': 1, 'reduce': 1, 'burden': 1, 'university': 1, 'centre': 1, 'residential': 1, 'liberal': 1, 'art': 1, 'jointly': 1, 'established': 1, 'Yale': 1, 'National': 1, 'Singapore': 1, 'utilises': 1, 'trained': 1, 'advisor': 1, 'triage': 1, 'manage': 1, 'lower-level': 1, 'including': 1, 'mild': 1, 'form': 1, 'stress': 1, 'social': 1, 'anxiety': 1, 'Staff': 1, 'complete': 1, 'in-house': 1, 'organised': 1, 'colleague': 1, 'extend': 1, 'external': 1, 'programme': 1, 'First': 1, 'Aid':

**Interpretasi**

**Topic 1 (Medical)**:
- Artikel ini memiliki total 13 tanda baca (titik) dan 10 tanda baca (koma).
- Kata-kata yang paling sering muncul dalam artikel ini adalah "student" dan "mental", masing-masing muncul sebanyak 8 kali.
- Kata-kata lain yang sering muncul termasuk "health", "support", "course", "faculty", "training", "resource", dan "college".
- Artikel ini membahas tentang dukungan kesehatan mental di sebuah institusi universitas, dengan menyoroti pelatihan staf dan dosen dalam mengelola masalah kesehatan mental mahasiswa.

**Topic 2 (Sport)**:
- Artikel ini memiliki total 21 tanda baca (koma) dan 13 tanda baca (titik).
- Kata-kata yang paling sering muncul dalam artikel ini adalah "Prawira" dan "Hong", masing-masing muncul sebanyak 4 kali.
- Kata-kata lain yang sering muncul termasuk "Basketball", "Eastern", "Round", "Asia", "Saputera", "Pelita", "Jaya", dan "Gist".
- Artikel ini membahas tentang pertandingan bola basket antara Prawira Harum Bandung dan Hong Kong Eastern Basketball Team.

**Topic 3 (Finance)**:
- Artikel ini memiliki total 14 tanda baca (titik) dan 12 tanda baca (koma).
- Kata-kata yang paling sering muncul dalam artikel ini adalah "tax" dan "rate", masing-masing muncul sebanyak 9 kali.
- Kata-kata lain yang sering muncul termasuk "proposal", "income", "capital", "gain", "Biden", "investment", dan "taxable".
- Artikel ini membahas tentang proposal peningkatan pajak atas capital gain yang diajukan oleh Presiden Joe Biden.

In [ ]:
# Vektor sebelum ditambahkan corpus_text
cv = CountVectorizer()
bow_vectors = cv.fit_transform(corpus_texts).toarray()
print(bow_vectors)
print(len(bow_vectors[0]))

[[0 0 0 ... 0 0 0]
 [1 0 1 ... 0 0 1]
 [0 2 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 1 ... 1 1 0]
 [0 0 0 ... 0 0 0]]
883


#### Klasifikasi Teks

##### Maximum similarity

Klasifikasi dilakukan berdasarkan maximum similarity, yaitu mengidentifikasi kategori atau label yang paling mirip dengan teks input berdasarkan kesamaannya dengan teks-teks yang sudah ada dalam data training.

###### Modeling

In [ ]:
# A new text to be classified based on topic
query_text = "Cristiano Ronaldo came off the bench to earn Manchester United a hard-fought 2-1 victory at Everton in the Premier League on Sunday, taking his career goal tally to 700 in the process. Just as United did last weekend in their derby mauling at the hands of local rivals Manchester City, they again found themselves behind early on at Goodison Park after Alex Iwobi curled a sublime strike into the net from 20 metres."
query_words = word_tokenize(query_text)
query_words_clean = [lemmatizer.lemmatize(word) for word in query_words if word not in set(stopwords.words('english'))]
query_words_corpus = [word for word in query_words_clean if word in set(bow_keys)]
query_text_corpus = ' '.join(query_words_corpus)
corpus_texts.append(query_text_corpus)

# Vektor setelah ditambahkan corpus_text
cv = CountVectorizer()
bow_vectors = cv.fit_transform(corpus_texts).toarray()
print(bow_vectors)
print(len(bow_vectors[0]))

[[0 0 0 ... 0 0 0]
 [1 0 1 ... 0 0 1]
 [0 2 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
883


Berikut merupakan penjelasan dari syntax di atas:
1. Yang pertama kali dilakukan adalah mendefinisikan teks yang akan diklasifikasikan

2. Melakukan tokenisasi dan preprocessing data

    a. Teks dari query_text di-tokenisasi menjadi kata-kata menggunakan word_tokenize.

    b. Setiap kata dalam teks di-lemmatisasi menggunakan lemmatizer.lemmatize(word) untuk mengubah kata-kata ke bentuk dasar.

    c. Kata-kata yang bukan stop words (kata-kata umum yang tidak memberikan banyak informasi seperti "the", "is", "and", dsb.) disimpan dalam variabel query_words_clean.

3. Kemudian dilanjut dengan seleksi kata-kata untuk corpus dan ditambahkan ke dalam corpus

    Kata-kata yang ada dalam query_words_clean dan juga ada dalam bow_keys (bag-of-words keys yang sudah dibersihkan sebelumnya) disimpan dalam variabel query_words_corpus.
    Kata-kata tersebut digabungkan kembali menjadi teks menggunakan join dan disimpan dalam variabel query_text_corpus.

4. Yang terakhir adalah membuat representasi vector dengan CountVectorizer

    a. Modul CountVectorizer dari scikit-learn digunakan untuk membuat representasi vektor dari teks.

    b. Objek cv merupakan objek CountVectorizer yang akan digunakan untuk mengubah teks menjadi vektor.

    c. cv.fit_transform(corpus_texts).toarray() digunakan untuk mengubah teks dari corpus_texts menjadi representasi vektor menggunakan metode bag-of-words.

    d. Hasilnya adalah bow_vectors, yang merupakan array numpy yang berisi vektor representasi dari teks-teks dalam corpus_texts.


**Interpretasi**

Dari output yang diperoleh, kita memiliki matriks yang mewakili Bag of Words (BoW) dari teks yang disertakan. Matriks tersebut memiliki 883 kolom, yang menunjukkan bahwa terdapat 883 kata unik dalam seluruh korpus teks, termasuk teks tambahan yang ingin diklasifikasikan.

Matriks tersebut memiliki 7 baris, yang masing-masing mewakili satu topik dalam korpus teks. Berikut adalah interpretasi baris-baris tersebut:

- **Baris 0**: Mewakili BoW untuk "Topic 1" yang merupakan topik tentang kesehatan mental di institusi universitas.
- **Baris 1**: Mewakili BoW untuk "Topic 2" yang merupakan topik tentang pertandingan bola basket antara tim-tim tertentu.
- **Baris 2**: Mewakili BoW untuk "Topic 3" yang merupakan topik tentang proposal peningkatan pajak atas capital gain yang diajukan oleh Presiden Joe Biden.
- **Baris 3-5**: Mewakili BoW untuk "Topic 4" sampai dengan "Topic 6" yang merupakan topik artikel untuk data test menggunakan model selain metode maximum similarity.
- **Baris 6**: Mewakili BoW untuk teks tambahan yang ingin diklasifikasikan. Dalam teks ini, tidak ada kata yang muncul, sehingga semua nilai dalam baris ini adalah 0.

Dengan menggunakan matriks BoW tersebut, kita dapat melanjutkan dengan tahap klasifikasi teks tambahan tersebut ke dalam salah satu topik berdasarkan kesamaannya dengan matriks BoW yang sudah ada.

In [ ]:
# Normalize the BoW Vectors
bow_texts_norm = []
for bow in bow_vectors:
  length = (sum(i*i for i in bow)) ** 0.5
  bow_norm = bow/length
  bow_texts_norm.append(bow_norm)

# Compute similarity using dot product
similarity_vector = []
bow_norm_query = bow_texts_norm[-1]
for bow in bow_texts_norm[:3]:
  similarity_vector.append(sum(i*j for i,j in zip(bow, bow_norm_query)))
print(similarity_vector)

# Find the highest similarity
id_max_sim = similarity_vector.index(max(similarity_vector))
if(id_max_sim == 0):
  print("The query text is classified as: Medical topic")
elif(id_max_sim == 1):
  print("The query text is classified as: Sport topic")
elif(id_max_sim == 2):
  print("The query text is classified as: Finance topic")

[0.0, 0.07286792513358903, 0.039461089893359204]
The query text is classified as: Sport topic


**Interpretasi**

Dalam output, diberikan nilai similaritas antara query text dengan setiap kelas, diikuti dengan kelas yang memiliki nilai similaritas tertinggi. Dalam contoh tersebut, nilai similaritas antara query text dengan kelas Medical, Sport, dan Finance berturut-turut adalah 0.0, 0.07286, dan 0.03946. Karena nilai similaritas tertinggi terjadi dengan kelas Sport, maka query text tersebut diklasifikasikan sebagai kelas Sport.

###### Metric Evaluation

In [ ]:
import numpy as np
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

true_label = "Sport"
predicted_label = "Sport"  # Sesuai dengan output dari kode
accuracy = accuracy_score([true_label], [predicted_label])
precision = precision_score([true_label], [predicted_label], average='weighted')
recall = recall_score([true_label], [predicted_label], average='weighted')
f1 = f1_score([true_label], [predicted_label], average='weighted')
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)


Accuracy: 1.0
Precision: 1.0
Recall: 1.0
F1-score: 1.0


Pada kasus klasifikasi teks, diperlukan evaluasi seberapa mirip teks (true positive) tersebut terhadap data yang sudah di-train sehingga pada kasus ini akan dipilih metric evaluation **akurasi**, untuk memberikan gambaran tentang seberapa sering model membuat prediksi yang benar secara keseluruhan.

Syntax di atas adalah sntax untuk menghitung metrik evaluasi hasil dari klasifikasi bag of words yang memberikan output Sport dari data testing, seperti akurasi (accuracy), presisi (precision), recall, dan F1-score. Berikut adalah penjelasannya:

1. **Accuracy (Akurasi):**
   - Akurasi mengukur seberapa tepat model dalam mengklasifikasikan data. Nilai akurasi berkisar dari 0 hingga 1, di mana 1 menunjukkan prediksi yang sempurna dan 0 menunjukkan prediksi yang buruk.
   - Dalam contoh tersebut, hasilnya adalah 1.0, yang berarti model melakukan prediksi dengan sempurna, atau dalam kasus ini, prediksi label yang sama dengan label sebenarnya (karena hasil output dari syntax klasifikasi di atas didapapatkan hasil label Sport)

2. **Precision (Presisi):**
   - Presisi mengukur seberapa andal model dalam mengidentifikasi instance positif dari semua instance yang diprediksi sebagai positif.
   - Hasilnya juga adalah 1.0, yang menunjukkan bahwa dari semua instance yang diprediksi sebagai "Sport", semuanya benar-benar merupakan "Sport". Ini menunjukkan tingkat keandalan model dalam mengklasifikasikan label yang positif.

3. **Recall (Recall atau Sensitivitas):**
   - Recall mengukur seberapa baik model dapat mengidentifikasi instance positif dari semua instance yang sebenarnya positif.
   - Seperti akurasi dan presisi, recall juga memiliki nilai 1.0 dalam contoh ini, menunjukkan bahwa model berhasil mengidentifikasi semua instance yang sebenarnya "Sport".

4. **F1-score (Skor F1):**
   - F1-score adalah rata-rata harmonik dari presisi dan recall. Metric ini berguna saat kita ingin memperhitungkan baik false positives maupun false negatives.
   - Hasilnya juga adalah 1.0, menunjukkan bahwa model memiliki keseimbangan yang baik antara presisi dan recall dalam kasus ini.

Secara keseluruhan, hasil syntax tersebut menunjukkan bahwa model melakukan dengan sangat baik dalam mengklasifikasikan teks yang diberikan sebagai "Sport", dengan nilai **akurasi**, presisi, recall, dan F1-score mencapai 1.0. Alasan pemberian 1 label (Sport) pada true_label dan predicted_label dikarenakan data testing yang diinput hanya berupa 1 teks, berbeda kasus jika data testing lebih dari 1 maka true_label akan semakin bertambah.

##### Decision Tree, Random Forest, Logistic Regression, & SVM

###### Modeling

In [ ]:
# Import Library
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
import warnings
warnings.filterwarnings("ignore")

In [ ]:
# Split index menjadi training, test, dan predict data
X_train = bow_texts_norm[0:3]
X_test = bow_texts_norm[3:6]
text_predict = bow_texts_norm[6]

# List string topic train & data
y_train = ['Medical', 'Sport', 'Finance']
y_test = ['Medical', 'Sport', 'Finance']

# Fit model train, y train
# Membuat model
dt = DecisionTreeClassifier()
rf = RandomForestClassifier()
lr = LogisticRegression()
svm = svm.SVC()

# List model
models = [dt, rf, lr, svm]
model_names = ['Decision Tree', 'Random Forest', 'Logistic Regression', 'SVM']

# Dictionary untuk menyimpan hasil prediksi
y_preds = {}

# Melakukan fit training dan prediksi untuk setiap model
for model, name in zip(models, model_names):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    y_preds[name] = y_pred

###### Metric Evaluation

In [ ]:
# Metric Evaluation
# Membuat classification report
print('### Testing Results Decision Tree Model ###\n')
print(classification_report(y_test, y_preds['Decision Tree']), '\n')

print('### Testing Results Random Forest Model ###\n')
print(classification_report(y_test, y_preds['Random Forest']), '\n')

print('### Testing Results Logistic Regression Model ###\n')
print(classification_report(y_test, y_preds['Logistic Regression']), '\n')

print('### Testing Results SVM Model ###\n')
print(classification_report(y_test, y_preds['SVM']), '\n')

### Testing Results Decision Tree Model ###

              precision    recall  f1-score   support

     Finance       0.50      1.00      0.67         1
     Medical       0.00      0.00      0.00         1
       Sport       1.00      1.00      1.00         1

    accuracy                           0.67         3
   macro avg       0.50      0.67      0.56         3
weighted avg       0.50      0.67      0.56         3
 

### Testing Results Random Forest Model ###

              precision    recall  f1-score   support

     Finance       0.50      1.00      0.67         1
     Medical       0.00      0.00      0.00         1
       Sport       1.00      1.00      1.00         1

    accuracy                           0.67         3
   macro avg       0.50      0.67      0.56         3
weighted avg       0.50      0.67      0.56         3
 

### Testing Results Logistic Regression Model ###

              precision    recall  f1-score   support

     Finance       1.00      1.00     

Pada kasus klasifikasi teks, diperlukan evaluasi seberapa mirip teks (true positive) tersebut terhadap data yang sudah di-train sehingga pada kasus ini akan dipilih metric evaluation **akurasi**, untuk memberikan gambaran tentang seberapa sering model membuat prediksi yang benar secara keseluruhan.

**Interpretasi**
* Akurasi yang dihasilkan oleh model Decision Tree adalah sebesar 67%
* Akurasi yang dihasilkan oleh model Random Forest adalah sebesar 67%
* Akurasi yang dihasilkan oleh model Logistic Regression adalah sebesar 100%
* Akurasi yang dihasilkan oleh model SVM adalah sebesar 100%

Artinya, dari ketiga model, yang paling baik modelnya adalah Logistic Regression dan SVM. Oleh karena itu, akan dipilih model Logistic Regression dan SVM untuk memprediksi text baru yang telah disiapkan.

###### New text prediction

In [ ]:
# Logistic Regression
lr.predict([text_predict])

array(['Sport'], dtype='<U7')

In [ ]:
# SVM
svm.predict([text_predict])

array(['Sport'], dtype='<U7')

Dari kedua model, diperoleh bahwa teks yang diberikan terprediksi sebagai Sport topic